### PR TITLE
Add shunts loadflow validation

### PR DESCRIPTION
**Pull request by massimo-ferraro**

### CODE REVIEW

File: loadflow/loadflow-validation/src/main/java/com/powsybl/loadflow/validation/ShuntCompensatorsValidation.java [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+        Objects.requireNonNull(config);
+        Objects.requireNonNull(shuntsWriter);
+        LOGGER.info("Checking shunt compensators of network {}", network.getId());
+        return network.getShuntStream()
+                      .sorted(Comparator.comparing(ShuntCompensator::getId))
+                      .map(shunt -> checkShunts(shunt, config, shuntsWriter))
+                      .reduce(Boolean::logicalAnd)
+                      .orElse(true);
+    }
+
```
Comment (mathbagu) [inline loadflow/loadflow-validation/src/main/java/com/powsybl/loadflow/validation/ShuntCompensatorsValidation.java]: These 3 methods are duplicated in all the *Validation classes. I think we have to do a refactor to avoid the duplication. It should be done in a different PR

File: loadflow/loadflow-validation/src/main/java/com/powsybl/loadflow/validation/ShuntCompensatorsValidation.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+        Objects.requireNonNull(config);
+        Objects.requireNonNull(shuntsWriter);
+
+        boolean validated = true;
+        float p = shunt.getTerminal().getP();
+        float q = shunt.getTerminal().getQ();
+        int currentSectionCount = shunt.getCurrentSectionCount();
+        int maximumSectionCount = shunt.getMaximumSectionCount();
+        float bPerSection = shunt.getbPerSection();
+        float qMax = shunt.getProperties().containsKey("qMax") ? Float.parseFloat((String) shunt.getProperties().get("qMax")) : Float.NaN;
```
Comment (mathbagu) [inline loadflow/loadflow-validation/src/main/java/com/powsybl/loadflow/validation/ShuntCompensatorsValidation.java]: It’s not a good idea to use this property. It exists only for a bijective import/export from and to the RTE study tool

File: loadflow/loadflow-validation/src/main/java/com/powsybl/loadflow/validation/ShuntCompensatorsValidation.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+        float nominalV = shunt.getTerminal().getVoltageLevel().getNominalV();
+        Bus bus = shunt.getTerminal().getBusView().getBus();
+        if (bus != null && !Float.isNaN(bus.getV())) {
+            float v = bus.getV();
+            return checkShunts(shunt.getId(), p, q, currentSectionCount, maximumSectionCount, bPerSection, v, qMax, nominalV, config, shuntsWriter);
+        } else if (!Float.isNaN(q) && q != 0) { // if the shunt is disconnected then either “q” is not defined or “q” is 0
```
Comment (mathbagu) [inline loadflow/loadflow-validation/src/main/java/com/powsybl/loadflow/validation/ShuntCompensatorsValidation.java]: If the shunt is disconnected, q should be NaN. 0 seems an invalid value

File: loadflow/loadflow-validation/src/main/java/com/powsybl/loadflow/validation/ShuntCompensatorsValidation.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    public static boolean checkShunts(String id, float p, float q, int currentSectionCount, int maximumSectionCount, float bPerSection,
+                                      float v, float qMax, float nominalV, ValidationConfig config, ValidationWriter shuntsWriter) {
+        boolean validated = true;
+        // “p” is always 0 or NaN
+        if (!Float.isNaN(p) && p != 0) {
```
Comment (mathbagu) [inline loadflow/loadflow-validation/src/main/java/com/powsybl/loadflow/validation/ShuntCompensatorsValidation.java]: Should be always NaN

File: loadflow/loadflow-validation/src/main/java/com/powsybl/loadflow/validation/ShuntCompensatorsValidation.java [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+            LOGGER.warn("{} {}: {}: P={}", ValidationType.SHUNTS, ValidationUtils.VALIDATION_ERROR, id, p);
+            validated = false;
+        }
+        // 0 lower or equal “currentSectionCount” lower or equal “maximumSectionCount”
+        if (currentSectionCount < 0 || currentSectionCount > maximumSectionCount) {
```
Comment (mathbagu) [inline loadflow/loadflow-validation/src/main/java/com/powsybl/loadflow/validation/ShuntCompensatorsValidation.java]: I think it’s already checked in the ShuntCompensatorAdder and in the setter of ShuntCompensator

File: loadflow/loadflow-validation/src/main/java/com/powsybl/loadflow/validation/ValidationTool.java [hunk 6]
```diff
@@ -86,7 +86,7 @@ public Options getOptions() {
                     .argName("VALIDATION_WRITER")
                     .build());
             options.addOption(Option.builder().longOpt(TYPES)
-                    .desc("validation types (FLOWS/GENERATORS/BUSES/SVCS) to run, all of them if the option if not specified")
+                    .desc("validation types (FLOWS/GENERATORS/BUSES/SVCS/SHUNTS) to run, all of them if the option if not specified")
```
Comment (mathbagu) [inline loadflow/loadflow-validation/src/main/java/com/powsybl/loadflow/validation/ValidationTool.java]: You should use the ValidationType values to generate this list as it’s done in the other tools

File: loadflow/loadflow-validation/src/main/java/com/powsybl/loadflow/validation/ValidationTool.java [hunk 7]
```diff
@@ -151,6 +151,7 @@ public void run(CommandLine line, ToolRunningContext context) throws Exception {
                 context.getErrorStream().println("Error validating load-flow results of network " + network.getId()
                                                  + " - validation type: " + validationType
                                                  + " - error: " + e.getMessage());
+                e.printStackTrace(context.getErrorStream());
```
Comment (mathbagu) [inline loadflow/loadflow-validation/src/main/java/com/powsybl/loadflow/validation/ValidationTool.java]: Is it necessary ?

File: loadflow/loadflow-validation/src/test/java/com/powsybl/loadflow/validation/io/AbstractValidationFormatterWriterTest.java [hunk 8]
```diff
@@ -191,4 +199,30 @@ protected void testSvcs(String svcsContent, boolean verbose) throws IOException
     }
 
     protected abstract ValidationWriter getSvcsValidationFormatterCsvWriter(TableFormatterConfig config, Writer writer, boolean verbose);
+
+    @Test
+    public void testShunts() throws Exception {
+        testShunts(getShuntsContent(), false);
+    }
+
+    protected abstract  String getShuntsContent();
```
Comment (mathbagu) [inline loadflow/loadflow-validation/src/test/java/com/powsybl/loadflow/validation/io/AbstractValidationFormatterWriterTest.java]: Remove this extra space between abstract and String

File: loadflow/loadflow-validation/src/main/java/com/powsybl/loadflow/validation/ShuntCompensatorsValidation.java (hunk 2 above)
Comment (massimo-ferraro) [reply loadflow/loadflow-validation/src/main/java/com/powsybl/loadflow/validation/ShuntCompensatorsValidation.java]: Is qMax available in any other way? Or should I skip the check on qMax (it is just for warning)?

Comment (mathbagu) [reply loadflow/loadflow-validation/src/main/java/com/powsybl/loadflow/validation/ShuntCompensatorsValidation.java]: You should skip the test and remove the 3 relevant lines in the checkShunt method

File: loadflow/loadflow-validation/src/main/java/com/powsybl/loadflow/validation/ShuntCompensatorsValidation.java (hunk 3 above)
Comment (massimo-ferraro) [reply loadflow/loadflow-validation/src/main/java/com/powsybl/loadflow/validation/ShuntCompensatorsValidation.java]: The validation can work both with or without running the loadflow. It seems that the Q value of a disconnected shunt is NaN after you run the loadflow, but is 0 after the import. Should I just check that the value is NaN?

Comment (mathbagu) [reply loadflow/loadflow-validation/src/main/java/com/powsybl/loadflow/validation/ShuntCompensatorsValidation.java]: We probably have an issue in the RTE tool which export all the IIDM files. The correct value should be NaN for disconnected injections.
Keep the test like this, we will do a fix in RTE tool and in the IIDM importer to fix these values one day
